src/porepy/numerics/ad/grid_operators.py

```diff
@@ -250,251 +268,417 @@ def __init__(
-                    # always as above.
-                    sz_mortar = intf.num_cells * dim
-                    if len(subdomains) == 0:
-                        # No subdomains provided. The total size is zero.
-                        sz_tot = 0
-                    else:
-                        # We need the number of rows for the primary projection matrix.
-                        # This equals total number of faces in the subdomains
-                        # times the dimension of the problem.
-                        sz_tot = sum([sd.num_cells for sd in subdomains]) * dim
-                    m2s, s2m = zero_matrices(sz_mortar, sz_tot)
-                    mortar_to_secondary_int.append(m2s)
-                    mortar_to_secondary_avg.append(m2s)
-                    secondary_to_mortar_int.append(s2m)
-                    secondary_to_mortar_avg.append(s2m)
+        # Store the list of subdomains and interfaces, as well as the mixed-dimensional
+        # grid. These will be needed to construct the projection matrices.
+        self._subdomains = subdomains
+        self._interfaces = interfaces
+        self._mdg = mdg
+
+        # Check if all interfaces are conforming. If so, we can use the same projection
+        # for intensive and extensive quantities. This must be done separately for
+        # mappings to primary and secondary grids. A more fine-grained check would be to
+        # consider each interface separately, but this does not seem worthwhile given
+        # the current usage of non-conforming grids.
+        is_conforming_primary = True
+        is_conforming_secondary = True
 
+        for intf in interfaces:
+            # Check the data of projections for both intensive and extensive quantities.
+            # If both have essentially unit values, the interface is conforming.
+            for proj in [intf.mortar_to_primary_int(), intf.mortar_to_primary_avg()]:
+                if not np.allclose(proj.data, 1, atol=1e-10):
+                    is_conforming_primary = False
+            for proj in [
+                intf.mortar_to_secondary_int(),
+                intf.mortar_to_secondary_avg(),
+            ]:
+                if not np.allclose(proj.data, 1, atol=1e-10):
+                    is_conforming_secondary = False
+
+        self._is_conforming_primary = is_conforming_primary
+        self._is_conforming_secondary = is_conforming_secondary
+
+        # Storage for the projection matrices. These will be constructed lazily, when
+        # the projection is requested, and then stored for later use. Depending on
+        # whether the interfaces are conforming or not, we need to store different
+        # projections.
+        if self._is_conforming_primary:
+            # Mapping for intensive and extensive quantities are the same, we can get
+            # away with storing only one set of projections.
+            self._mortar_to_primary: Optional[SparseArray] = None
```

Comment:
I recommend a brief explanation of the abbreviations "int" and "avg" before "int" is connected to "intensive" here. Possibly order "intensive and extensive" in the doc in the same way as used in the code.

Reply:
The class docstring could refer to the tutorial

Reply:
Agree, @IvarStefansson. Will do.